lib/logstash/outputs/elasticsearch/http_client.rb

```diff
@@ -38,23 +40,45 @@ def get_version
     def bulk(actions)
       @action_count ||= 0
       @action_count += actions.size
-      
+
       return if actions.empty?
-      bulk_body = actions.collect do |action, args, source|
+
+
+      bulk_actions = actions.collect do |action, args, source|
         args, source = update_action_builder(args, source) if action == 'update'
 
         if source && action != 'delete'
           next [ { action => args }, source ]
         else
           next { action => args }
         end
-      end.
-      flatten.
-      reduce("") do |acc,line|
-        acc << LogStash::Json.dump(line)
-        acc << "\n"
       end
 
+      bulk_body = ""
+      bulk_actions.each do |action|
+        as_json = action.is_a?(Array) ?
+                    action.map {|line| LogStash::Json.dump(line)}.join("\n") :
+                    LogStash::Json.dump(action)
+
+        if as_json.size > @max_doc_bytes
+          @logger.warn("Document too large for ES output! Dropping",
+                        :max_doc_megabytes => @options[:max_doc_megabytes],
+                        :bulk_action_size => as_json.size)
+          next
+        end
+
+        if (bulk_body.size + as_json.size) > @target_bulk_bytes
```

Comment:
Do we have to call `bulk_body.size` every time? Maybe we memoize this size, for the in-flight `bulk_batch`? As in, keep a running total of the current batch size. You only need to calculate size for current doc every time, and then add it to `bulk_batch_size`. Then reset `bulk_batch_size` when the request is complete or when an error happens. WDYT?